shell sort

sreshtech opened the pull request:
<!-- For completed items, change [ ] to [x] -->
## Pull Request Checkboxes and descriptions

### Description
Shell Sort is an efficient and adaptive sorting algorithm that is an extension of the Insertion Sort algorithm. It works by dividing the array into smaller subarrays and then sorting these subarrays using the Insertion Sort method. The key idea behind Shell Sort is to compare and swap elements that are far apart, gradually reducing the gap between elements until the entire array is sorted.
Basic Steps:
1.Choose a gap sequence (commonly, it starts with n/2 where n is the length of the array) to determine the initial gap.
2.Divide the array into subarrays of elements separated by the chosen gap.
3.Apply Insertion Sort to each subarray independently to sort the elements within the subarray.
4.Reduce the gap between elements and repeat steps 2 and 3 until the gap becomes 1.
5.Finally, apply Insertion Sort with a gap of 1 to perform the final sorting, which is usually efficient because the array has been partially sorted by the previous steps.
Shell Sort is known for its adaptive nature, as the choice of gap sequence can significantly affect its performance. It has a time complexity that depends on the gap sequence used but often exhibits a better average-case performance compared to the simple Insertion Sort.

### Wiki link
https://en.wikipedia.org/wiki/Shellsort


### Changes Made
[I have made a valuable enhancement to the repository by incorporating the Shell sort algorithm into the sorting module. This significant addition greatly enhances the repository's sorting capabilities, contributing a fundamental algorithm that was previously absent from the sorting section. The inclusion of the Shell sort algorithm will empower users with an efficient and versatile sorting solution, further enriching the functionality and versatility of this repository]

### Screenshots (if applicable)
[Include any relevant screenshots or images here.]

- [x] Fixed Bug/issues
- [x] Added algo/questions_answers/docs
- [x] This pull request is all my own work -- I have not plagiarized it.
- [x] All filenames are in PascalCase.
- [x] All functions and variable names follow Java naming conventions.
- [x] All new algorithms have a URL in their comments that points to Wikipedia or other similar explanations.

rohan472000 commented:
@sreshtech kindly update the PR description above by adding changes briefly and ticking checkboxes correctly.

rohan472000 commented:
@sreshtech I can't see any file in PR....

rohan472000 reviewed (approved):
Provide algo description briefly and give a wiki link also

sreshtech commented:
> @sreshtech I can't see any file in PR....

I apologize for inadvertently deleting the file while making other necessary changes to the repository. i resolved the issue .thank you .

sreshtech commented:
> Provide algo description briefly and give a wiki link also
Resolved .thank you !

rohan472000 commented:
I can't see any Wikipedia link or description

sreshtech commented:
> I can't see any Wikipedia link or description 
sorry can you please tell where to add this because i edited it above.please help 
<!-- For completed items, change [ ] to [x] -->
## Pull Request Checkboxes and descriptions

### Description
Shell Sort is an efficient and adaptive sorting algorithm that is an extension of the Insertion Sort algorithm. It works by dividing the array into smaller subarrays and then sorting these subarrays using the Insertion Sort method. The key idea behind Shell Sort is to compare and swap elements that are far apart, gradually reducing the gap between elements until the entire array is sorted.
Basic Steps:
1.Choose a gap sequence (commonly, it starts with n/2 where n is the length of the array) to determine the initial gap.
2.Divide the array into subarrays of elements separated by the chosen gap.
3.Apply Insertion Sort to each subarray independently to sort the elements within the subarray.
4.Reduce the gap between elements and repeat steps 2 and 3 until the gap becomes 1.
5.Finally, apply Insertion Sort with a gap of 1 to perform the final sorting, which is usually efficient because the array has been partially sorted by the previous steps.
Shell Sort is known for its adaptive nature, as the choice of gap sequence can significantly affect its performance. It has a time complexity that depends on the gap sequence used but often exhibits a better average-case performance compared to the simple Insertion Sort.

### Wiki link
https://en.wikipedia.org/wiki/Shellsort


### Changes Made
[I have made a valuable enhancement to the repository by incorporating the Shell sort algorithm into the sorting module. This significant addition greatly enhances the repository's sorting capabilities, contributing a fundamental algorithm that was previously absent from the sorting section. The inclusion of the Shell sort algorithm will empower users with an efficient and versatile sorting solution, further enriching the functionality and versatility of this repository]

### Screenshots (if applicable)
[Include any relevant screenshots or images here.]

- [x] Fixed Bug/issues
- [x] Added algo/questions_answers/docs
- [x] This pull request is all my own work -- I have not plagiarized it.
- [x] All filenames are in PascalCase.
- [x] All functions and variable names follow Java naming conventions.
- [x] All new algorithms have a URL in their comments that points to Wikipedia or other similar explanations.
 
i edited it above

rohan472000 commented:
No..you get this wrong...see the last checkbox option in that you should include link in comments which means under .java file not in Par description.

rohan472000 commented on sorting/shell_sort.java:
```diff
@@ -1,3 +1,18 @@
+4.Reduce the gap between elements and repeat steps 2 and 3 until the gap becomes 1.
+5.Finally, apply Insertion Sort with a gap of 1 to perform the final sorting, which is usually efficient because the array has been partially sorted by the previous steps.
+Shell Sort is known for its adaptive nature, as the choice of gap sequence can significantly affect its performance. 
+It has a time complexity that depends on the gap sequence used but often exhibits a better average-case performance compared to the simple Insertion Sort.
```
```suggestion

```

sreshtech commented:
> No..you get this wrong...see the last checkbox option in that you should include link in comments which means under .java file not in Par description.

sorry i think i resolved it now. if i am still wrong please help.

rohan472000 commented on sorting/shell_sort.java:
```diff
@@ -1,3 +1,18 @@
+Shell Sort is an efficient and adaptive sorting algorithm that is an extension of the Insertion Sort algorithm.
+It works by dividing the array into smaller subarrays and then sorting these subarrays using the Insertion Sort method. 
+The key idea behind Shell Sort is to compare and swap elements that are far apart, gradually reducing the gap between elements until the entire array is sorted.
+Basic Steps:
```
```suggestion

```

rohan472000 commented on sorting/shell_sort.java:
```diff
@@ -1,3 +1,18 @@
+It works by dividing the array into smaller subarrays and then sorting these subarrays using the Insertion Sort method. 
+The key idea behind Shell Sort is to compare and swap elements that are far apart, gradually reducing the gap between elements until the entire array is sorted.
+Basic Steps:
+1.Choose a gap sequence (commonly, it starts with n/2 where n is the length of the array) to determine the initial gap.
```
```suggestion
```

rohan472000 commented on sorting/shell_sort.java:
```diff
@@ -1,3 +1,18 @@
+The key idea behind Shell Sort is to compare and swap elements that are far apart, gradually reducing the gap between elements until the entire array is sorted.
+Basic Steps:
+1.Choose a gap sequence (commonly, it starts with n/2 where n is the length of the array) to determine the initial gap.
+2.Divide the array into subarrays of elements separated by the chosen gap.
```
```suggestion

```

rohan472000 commented on sorting/shell_sort.java:
```diff
@@ -1,3 +1,18 @@
+Basic Steps:
+1.Choose a gap sequence (commonly, it starts with n/2 where n is the length of the array) to determine the initial gap.
+2.Divide the array into subarrays of elements separated by the chosen gap.
+3.Apply Insertion Sort to each subarray independently to sort the elements within the subarray.
```
```suggestion

```

rohan472000 commented on sorting/shell_sort.java:
```diff
@@ -1,3 +1,18 @@
+1.Choose a gap sequence (commonly, it starts with n/2 where n is the length of the array) to determine the initial gap.
+2.Divide the array into subarrays of elements separated by the chosen gap.
+3.Apply Insertion Sort to each subarray independently to sort the elements within the subarray.
+4.Reduce the gap between elements and repeat steps 2 and 3 until the gap becomes 1.
```
```suggestion

```

rohan472000 commented on sorting/shell_sort.java:
```diff
@@ -1,3 +1,18 @@
+2.Divide the array into subarrays of elements separated by the chosen gap.
+3.Apply Insertion Sort to each subarray independently to sort the elements within the subarray.
+4.Reduce the gap between elements and repeat steps 2 and 3 until the gap becomes 1.
+5.Finally, apply Insertion Sort with a gap of 1 to perform the final sorting, which is usually efficient because the array has been partially sorted by the previous steps.
```
```suggestion

```

rohan472000 commented on sorting/shell_sort.java:
```diff
@@ -1,3 +1,18 @@
+3.Apply Insertion Sort to each subarray independently to sort the elements within the subarray.
+4.Reduce the gap between elements and repeat steps 2 and 3 until the gap becomes 1.
+5.Finally, apply Insertion Sort with a gap of 1 to perform the final sorting, which is usually efficient because the array has been partially sorted by the previous steps.
+Shell Sort is known for its adaptive nature, as the choice of gap sequence can significantly affect its performance. 
```
```suggestion

```
